go/service/identify3.go

```diff
@@ -0,0 +1,245 @@
+package service
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/keybase/client/go/engine"
+	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/go-framed-msgpack-rpc/rpc"
+	"golang.org/x/net/context"
+	"stathat.com/c/ramcache"
+	"sync"
+	"time"
+)
+
+// identify3Session corresponds to a single screen showing a user profile.
+// It maps 1:1 with an Identify2GUIID, and is labeled as such. Here we'll keep
+// track of whatever context we need to pass across calls, and also the TrackToken
+// for the final result.
+type identify3Session struct {
+	sync.Mutex
+	created time.Time
+	id      keybase1.Identify3GUIID
+	outcome *libkb.IdentifyOutcome
+}
+
+func newIdentify3GUIID() (keybase1.Identify3GUIID, error) {
+	var b []byte
+	l := 12
+	b, err := libkb.RandBytes(l)
+	if err != nil {
+		return keybase1.Identify3GUIID(""), err
+	}
+	b[l-1] = 0x34
+	return keybase1.Identify3GUIID(hex.EncodeToString(b)), nil
+}
+
+func newIdentify3Session(mctx libkb.MetaContext) (*identify3Session, error) {
+	id, err := newIdentify3GUIID()
+	if err != nil {
+		return nil, err
+	}
+	ret := &identify3Session{
+		created: mctx.G().GetClock().Now(),
+		id:      id,
+	}
+	mctx.CDebugf("generated new identify3 session: %s", id)
+	return ret, nil
+}
+
+// identify3Handler handles one RPC in the identify3 flow
+type identify3Handler struct {
+	libkb.Contextified
+	state *identify3State
+	xp    rpc.Transporter
+}
+
+// identify3State keeps track of all active ID3 state across the whole app. It has
+// a cache that's periodically cleaned up.
+type identify3State struct {
+	sync.Mutex
+
+	// Table of keybase1.Identify3GUIID -> *identify3Session's
+	cache *ramcache.Ramcache
+}
+
+func newIdentify3State(g *libkb.GlobalContext) *identify3State {
+	cache := ramcache.New()
+	cache.MaxAge = 24 * time.Hour
+	return &identify3State{
+		cache: cache,
+	}
+}
+
+// get an identify3Session out of the cache, as keyed by a Identify3GUIID. Return
+// (nil, nil, nil) if not found. Return (nil, nil, Error) if there was an expected error.
+// Return (i, f, nil) if found, where i is the **locked** object, f is the unlocker function
+// that you are required to call if non-nil, and nil is the nil error.
+func (s *identify3State) get(key keybase1.Identify3GUIID) (ret *identify3Session, unlocker func(), err error) {
+	s.Lock()
+	defer s.Unlock()
+	return s.getLocked(key)
+}
```

Comment:
I think it's nice for `unlocker` to [always be callable](https://github.com/keybase/client/blob/master/go/stellar/global.go#L455). That way [the caller](https://github.com/keybase/client/blob/master/go/stellar/build.go#L37) can just `defer unlocker()` no matter what and not think about it too much.